### PR TITLE
Implement new await synchronous interface

### LINKED ISSUE #288: Improve the gem's public api by removing the bang operator in some methods
Original discussion https://github.com/meilisearch/meilisearch-ruby/pull/273/files#r779759981

Today we use `!`(bang) operator in some particular cases eg. `create_index!`, `add_documents!`, these methods have something in common, *they wait for the task completion*.

After v0.25.0 index creations are totally async, so they respond with a task object not an index anymore. Anyway, what is the real problem here is the usage of bang makes a lot of difference and the typing structure is close to the other methods.

You could type `create_index` and get a `task` as a response, or  
You could type `create_index!` and get an `index` as a response.

The proposal here is to remove this syntax in order to provide something more clear to prevent such mistakes like:

`create_index('movies', sync: true)`, `create_index('movies', wait: true)`, `create_index('movies', wait_for_completion: true)`, in a way that if we don’t pass such param we know that it will respond with default `task` object.

### LINKED ISSUE #499: A test asserts incorrect behavior
**Description**
A weird test that asserts opposite behavior from what actually happens.

**Expected behavior**
From what I understand from the test's description it is _supposed_ to test behavior when a primary key `A` was defined during `create_index` and you try to call `add_documents` with a primary key `B`. The real behavior in this case is that the task fails, but the test claims that the documents are added anyway. 

**Current behavior**
The test claims one thing and tests something completely different.

**Screenshots or Logs**
This is the test in question
```rb
# spec/meilisearch/index/documents_spec.rb
  context 'Impossible to update primary-key if already given during index creation' do
    let(:documents) do
      { id: 1, unique: 1, title: 'Le Rouge et le Noir' }
    end

    it 'adds the documents anyway' do
      task = index.add_documents(documents, 'unique')
      expect(task).to be_a(Hash)
      client.wait_for_task(task['taskUid'])
      expect(index.fetch_primary_key).to eq('unique')
      expect(index.documents['results'].count).to eq(1)
    end
  end
```

### LINKED ISSUE #500: Redundant tests in settings_spec
**Description**
A collection of tests in `spec/meilisearch/index/settings_spec` that test behavior that has already been tested. Edit: While this is not always a sign that the tests are unnecessary, removing them does not change the coverage at all.

**Screenshots or Logs**
The tests in question
1. Testing `#settings`, `#update_settings`, and `#reset_settings` on an index with a **specified primary key** (which does not change the behavior of those methods, all of which are tested [earlier in the file in almost the exact same way](https://github.com/meilisearch/meilisearch-ruby/blob/fb6bbc0bcf2a39dcd7da230a0ca8fe43947493f8/spec/meilisearch/index/settings_spec.rb#L37))
<details><summary>Details</summary>
<p>

```ruby
  context 'Index with primary-key' do
    let(:index) { client.index(uid) }

    before { client.create_index!(uid, primary_key: 'id') }

    it 'gets the default values of settings' do
      settings = index.settings
      expect(settings).to be_a(Hash)
      expect(settings.keys).to contain_exactly(*settings_keys)
      expect(settings['rankingRules']).to eq(default_ranking_rules)
      expect(settings['distinctAttribute']).to be_nil
      expect(settings['searchableAttributes']).to eq(default_searchable_attributes)
      expect(settings['displayedAttributes']).to eq(default_displayed_attributes)
      expect(settings['stopWords']).to eq([])
      expect(settings['synonyms']).to eq({})
    end

    it 'updates multiples settings at the same time' do
      task = index.update_settings(
        ranking_rules: ['title:asc', 'typo'],
        distinct_attribute: 'title'
      )

      expect(task['type']).to eq('settingsUpdate')
      client.wait_for_task(task['taskUid'])
      settings = index.settings
      expect(settings['rankingRules']).to eq(['title:asc', 'typo'])
      expect(settings['distinctAttribute']).to eq('title')
      expect(settings['stopWords']).to be_empty
    end

    it 'updates one setting without reset the others' do
      task = index.update_settings(stop_words: ['the'])

      expect(task['type']).to eq('settingsUpdate')
      client.wait_for_task(task['taskUid'])
      settings = index.settings
      expect(settings['rankingRules']).to eq(default_ranking_rules)
      expect(settings['distinctAttribute']).to be_nil
      expect(settings['stopWords']).to eq(['the'])
      expect(settings['synonyms']).to be_empty
    end

    it 'resets all settings' do
      task = index.update_settings(
        ranking_rules: ['title:asc', 'typo'],
        distinct_attribute: 'title',
        stop_words: ['the'],
        synonyms: {
          wow: ['world of warcraft']
        }
      )
      expect(task['type']).to eq('settingsUpdate')
      client.wait_for_task(task['taskUid'])

      task = index.reset_settings

      expect(task['type']).to eq('settingsUpdate')
      client.wait_for_task(task['taskUid'])

      settings = index.settings
      expect(settings['rankingRules']).to eq(default_ranking_rules)
      expect(settings['distinctAttribute']).to be_nil
      expect(settings['stopWords']).to be_empty
      expect(settings['synonyms']).to be_empty
    end
  end
```
</p>
</details> 

2. Testing document addition (which is tested in `spec/index/document_spec.rb`) and searchable/displayed attributes which are tested earlier in the file [here](https://github.com/meilisearch/meilisearch-ruby/blob/fb6bbc0bcf2a39dcd7da230a0ca8fe43947493f8/spec/meilisearch/index/settings_spec.rb#L201) and [here](https://github.com/meilisearch/meilisearch-ruby/blob/fb6bbc0bcf2a39dcd7da230a0ca8fe43947493f8/spec/meilisearch/index/settings_spec.rb#L248), respectively

<details><summary>Details</summary>
<p>

```ruby
  context 'Manipulation of searchable/displayed attributes with the primary-key' do
    let(:index) { client.index(random_uid) }

    it 'does not add document when there is no primary-key' do
      task = index.add_documents(title: 'Test')
      task = client.wait_for_task(task['taskUid'])

      expect(task.keys).to include('error')
      expect(task['error']['code']).to eq('index_primary_key_no_candidate_found')
    end

    it 'adds documents when there is a primary-key' do
      task = index.add_documents(objectId: 1, title: 'Test')

      client.wait_for_task(task['taskUid'])
      expect(index.documents['results'].count).to eq(1)
    end

    it 'resets searchable/displayed attributes' do
      task = index.update_displayed_attributes(['title', 'description'])
      client.wait_for_task(task['taskUid'])
      task = index.update_searchable_attributes(['title'])

      client.wait_for_task(task['taskUid'])

      task = index.reset_displayed_attributes

      client.wait_for_task(task['taskUid'])
      expect(index.task(task['taskUid'])['status']).to eq('succeeded')

      task = index.reset_searchable_attributes

      client.wait_for_task(task['taskUid'])
      expect(index.task(task['taskUid'])['status']).to eq('succeeded')

      expect(index.displayed_attributes).to eq(['*'])
      expect(index.searchable_attributes).to eq(['*'])
    end
  end
```
</p></details>

### PR DESCRIPTION
**Pull request by ellnix**
# Pull Request

## Related issue
Fixes #288

While refactoring, inadvertently fixes:
- Fixes #499
- Fixes #500

## PR checklist
Please check if your PR fulfills the following requirements:
- [X] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [X] Have you read the contributing guidelines?
- [X] Have you made sure that the title is accurate and descriptive of the changes?

## Changes Checklist
- [X] Add soft deprecation messages
- [X] Remove bang methods from specs
- [X] Add Task "model"
- [X] Ensure all tests pass with edits where necessary
- [X] Add tests to make sure that all bang methods actually provide deprecation warnings 
- [X] Return Task model on all methods that return tasks
  - [X] Return Task model from all settings setters and getters
- [ ] Update the documentation with new syntax (if applicable)

### CODE REVIEW

**Comment by ellnix** [comment]
@brunoocasali this is pretty much complete, please feel free to review when you have (a lot of) time.

**Comment by ellnix** [comment]
Rebased and force pushed.

**Review by brunoocasali — COMMENTED** [review]
Thanks a lot for your hard work here @ellnix amazing!

I am asking for some changes but it is nothing to be worried about!

**Review by brunoocasali — CHANGES_REQUESTED** [review]
Amazing work @ellnix I think we could make some small modifications, but I was impressed, very good work!

**Comment by ellnix** [comment]
@brunoocasali sorry for taking so long to get back to this. I addressed your feedback and rebased it to main (for the changes since you last reviewed it, start at commit [b141ed4](https://github.com/meilisearch/meilisearch-ruby/pull/498/commits/b141ed48198449090f93ede4d7281ae65b01960b)).

**Comment by brunoocasali** [comment]
let's see if the code coverage will still be broken now!

Also, can you squash the commits @ellnix ?

**Comment by ellnix** [comment]
> Also, can you squash the commits @ellnix ?

Squash them into a single commit or just fewer commits? How many commits would you consider ideal? The history might be a little bit hard to follow if only a few commits had +1800 -1200 changes.

File: lib/meilisearch/utils.rb
Comment (brunoocasali) [reply lib/meilisearch/utils.rb:79]: > I took the liberty to unify the warning logging with a Rails-like logger. 

Good for me!

`MeiliSearch.logger` is better indeed!

**Comment by brunoocasali** [comment]
> Squash them into a single commit or just fewer commits? How many commits would you consider ideal? The history might be a little bit hard to follow if only a few commits had +1800 -1200 changes.

A few commits would be great. I don't have a number for that, it is more a feeling than anything. If you squash them based on the group like, you did a lot of the same but in multiple files, so in my mind you would have a single commit of that "theme".

**Comment by ellnix** [comment]
@brunoocasali should be ready for merge, unless you want me to squash it down further.

File: spec/meilisearch/models/task_spec.rb
Comment (brunoocasali) [inline spec/meilisearch/models/task_spec.rb:69]: When I mention the xUnit I mean this:

```suggestion
        allow(subject).to receive(:refresh)
        
        subject.enqueued?
        
        expect(subject).not_to have_received(:refresh)
```

**Review by brunoocasali — APPROVED** [review]
Looks good to me, let get this merged @ellnix

Thanks a lot for all the hard work in this repo @ellnix ❤️

**Comment by ellnix** [comment]
I'm not sure this will work, but addressing the conflict with a commit might dismiss the approval review.

bors merge

**Comment by ellnix** [comment]
Give me a second with this, apparently the git pull on my fork didn't go through properly and now everything is messed up.

**Comment by ellnix** [comment]
Well I fixed the git history and force pushed. Sorry about the issues but I don't think there was a way for me to resolve the main conflicts without automatically dismissing the approval review (even when I tried a merge rather than a rebase).

I looked it up, it seems to be a setting (https://github.com/isaacs/github/issues/783#issuecomment-320211882) but it does not look like it would be a good idea, since it would allow someone with approval to push arbitrary code and then choose to merge.

**Review by brunoocasali — APPROVED** [review]
LGTM!

**Comment by ellnix** [comment]
bors merge

**Pull request merged**
